### PR TITLE
Include grad3d example 

### PR DESCRIPTION
**Pull request by rfj82982**
Add example to compute the gradient of a field using global transpose routines. The example show a full example of a code using the 2DECOMP&FFT library.

### CODE REVIEW

File: examples/grad3d/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Halo test
```
Comment (mathrack) [inline examples/grad3d/README.md:1]: Gradient test ?

File: examples/grad3d/grad3d.f90 [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+   !=====================================================================
+   subroutine allocate_var()
+
+      use decomp_2d
+
+      implicit none
+
+      logical, parameter :: glob = .false.
+
+      ! initialise u,v,w with random numbers in X-pencil
```
Comment (mathrack) [inline examples/grad3d/grad3d.f90:137]: ! initialise phi with 3D harmonic function

File: examples/grad3d/grad3d.f90 [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+      integer :: xs1, xs2, xs3
+
+      xe1 = xsize(1)
+      xe2 = xsize(2)
+      xe3 = xsize(3)
+      xs1 = xstart(1)
+      xs2 = xstart(2)
+      xs3 = xstart(3)
+
+      ! Initial velocity
```
Comment (mathrack) [inline examples/grad3d/grad3d.f90:174]: velocity => scalar

File: examples/grad3d/grad3d.f90 [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+         df(i,j,k) = coeff*(ff(i,j,k+1)-ff(i,j,k-1))
+      enddo
+      do concurrent (j=1:ny, i=1:nx)
+         df(i,j,nz) = coeff*(ff(i,j,1)-ff(i,j,nz-1))
+      enddo
+      !$acc end kernels
+
+   end subroutine derz 
+   !=====================================================================
+   ! Calculate divergence using halo-cell exchange (data in X-pencil)
```
Comment (mathrack) [inline examples/grad3d/grad3d.f90:316]: Test derivatives against analytical solution (data in X-pencil)

File: examples/grad3d/grad3d.f90 [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+      integer :: xs1, xs2, xs3
+
+      xe1 = xsize(1)
+      xe2 = xsize(2)
+      xe3 = xsize(3)
+      xs1 = xstart(1)
+      xs2 = xstart(2)
+      xs3 = xstart(3)
+
+      ! Initial velocity
```
Comment (mathrack) [inline examples/grad3d/grad3d.f90:342]: Compute the error against the analytical solution

File: examples/grad3d/grad3d.f90 [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
```
Comment (mathrack) [inline examples/grad3d/grad3d.f90:396]: Compute the error against the analytical solution

File: examples/grad3d/grad3d.f90 [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
```
Comment (mathrack) [inline examples/grad3d/grad3d.f90:451]: Compute the error against the analytical solution

**Review by mathrack — CHANGES_REQUESTED** [review]
Some minor comments

File: examples/grad3d/grad3d.f90 (hunk 2 above)
Comment (rfj82982) [reply examples/grad3d/grad3d.f90:137]: Change with `! Allocate main variables in X-pencil`

File: examples/grad3d/grad3d.f90 (hunk 4 above)
Comment (rfj82982) [reply examples/grad3d/grad3d.f90:316]: Done

File: examples/grad3d/grad3d.f90 (hunk 5 above)
Comment (rfj82982) [reply examples/grad3d/grad3d.f90:342]: done

File: examples/grad3d/grad3d.f90 [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+
+      ! Compute the error against analytical solution
+      !$acc parallel loop default(present) reduction(+:error)
+      do k=1,xe3
+        do j=1,xe2
+          do i=1,xe1
+             z = (k + xs3 - 2) * dz
+             y = (j + xs2 - 2) * dy
+             x = (i + xs1 - 2) * dx
+             dphi = -2._mytype * twopi * sin(twopi * (x / lx)) * cos(twopi * (y / ly)) * sin(twopi * (z / lz))
```
Comment (pbartholomew08) [inline examples/grad3d/grad3d.f90:350]: Shouldn't this be `-2 * (twopi / lx) * ...`?

Comment (pbartholomew08) [reply examples/grad3d/grad3d.f90:350]: and similar for analytical Y and Z derivatives?

**Review by pbartholomew08 — COMMENTED** [review]
I think there is a minor error in the analytical solutions (or coffee hasn't kicked in yet!)

File: examples/grad3d/grad3d.f90 (hunk 8 above)
Comment (rfj82982) [reply examples/grad3d/grad3d.f90:350]: yes this is a mistake. In this case is ok since lx=ly=lz=1.0 but i need to add it

Comment (rfj82982) [reply examples/grad3d/grad3d.f90:350]: I forgot to put back in `l` when i did the derivatives

Comment (pbartholomew08) [reply examples/grad3d/grad3d.f90:350]: > yes this is a mistake. In this case is ok since lx=ly=lz=1.0 but i need to add it

Ah yes, good point :)

Comment (rfj82982) [reply examples/grad3d/grad3d.f90]: I have corrected the analytical formula. Just to sort the ADIOS2

File: examples/grad3d/grad3d.f90 (hunk 6 above)
Comment (rfj82982) [reply examples/grad3d/grad3d.f90]: done

File: examples/grad3d/grad3d.f90 (hunk 3 above)
Comment (rfj82982) [reply examples/grad3d/grad3d.f90:174]: done

**Pull request merged**